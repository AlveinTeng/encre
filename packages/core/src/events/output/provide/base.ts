import { GenerationChunk } from './generation.js';

/**
 * Interface that defines the source provider for providing readable
 * source.
 */
export interface SourceProvider<T = unknown> {
  source: T;

  generation: GenerationChunk;
}

export abstract class BaseSourceProvider<T = unknown> implements SourceProvider<T> {
  source: T;

  generation: GenerationChunk;

  abstract replace(source: T): void;

  abstract add(source: T): void;

  abstract provide(): GenerationChunk['output'];
}