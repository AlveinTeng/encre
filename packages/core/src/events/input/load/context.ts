export type ContextMetadata = Record<string, unknown>;

/**
 * Interface for interacting with a readable context.
 */
export interface ContextInput<ContextMetadata = Record<string, unknown>> {
  pageContent: string;

  metadata?: ContextMetadata;
}

/**
 * Class for constructing a readable context with content in one page.
 */
export class Context<ContextMetadata = Record<string, unknown>>
  implements ContextInput<ContextMetadata>
{
  pageContent: string;

  metadata: ContextMetadata;

  constructor(fields: ContextInput<ContextMetadata>) {
    this.pageContent = fields.pageContent
      ? fields.pageContent.toString()
      : this.pageContent;
    this.metadata = fields.metadata ?? ({} as ContextMetadata);
  }
}
