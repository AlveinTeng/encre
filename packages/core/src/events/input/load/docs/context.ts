/**
 * Interface for interacting with a readable context.
 */
export interface ContextInput<Metadata = Record<string, unknown>> {
  pageContent: string;

  metadata?: Metadata;
}

/**
 * Class for constructing a readable context with content in one page.
 */
export class Context<Metadata = Record<string, unknown>>
  implements ContextInput<Metadata>
{
  pageContent: string;

  metadata: Metadata;

  constructor(fields: ContextInput<Metadata>) {
    this.pageContent = fields.pageContent
      ? fields.pageContent.toString()
      : this.pageContent;
    this.metadata = fields.metadata ?? ({} as Metadata);
  }
}

export type ContextLike = Context | Context[] | string;
