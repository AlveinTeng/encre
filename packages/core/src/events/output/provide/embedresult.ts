/**
 * Contains all relevant information returned by an LLM Embeddings.
 */
export type EmbedResult = {
  embedding: number[];

  /**
   * Metadata from the LLM embedding API
   */
  embedOutput?: Record<string, any>;
};