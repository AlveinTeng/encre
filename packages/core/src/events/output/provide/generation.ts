export interface Generation {
  /**
   * Generated output
   */
  output: unknown;

  /**
   * Raw generation information from the provider.
   * May include reason for finishing.
   */
  info?: Record<string, unknown>;
}

export type GenerationChunkField = Generation;

export abstract class GenerationChunk implements Generation {
  public output: unknown;

  public info?: Record<string, unknown>;

  constructor(fields: GenerationChunkField) {
    this.output = fields.output;
    this.info = fields.info;
  }

  abstract concat(chunk: GenerationChunk): GenerationChunk;
}