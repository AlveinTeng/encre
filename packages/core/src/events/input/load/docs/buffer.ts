import { type readFile } from 'node:fs/promises';
import { BaseLoader } from './base.js';
import { Context } from './context.js';

/**
 * Abstract class that extends the `BaseLoader` class. It is a
 * document loader that loads document from a buffer. The `load()`
 * method is implemented to read the buffer contents and metadata,
 * and then calls the `parse()` method to parse the buffer and
 * return the documents.
 */
export abstract class BufferLoader<
  CallInput extends string | Blob = string | Blob,
> extends BaseLoader<CallInput> {

  _isSerializable = true;

  static _name(): string {
    return 'buffer';    
  }

  _loaderType(): string {
    return 'buffer';
  }

  /**
   * Abstract method that parse the buffer and return the documents.
   * @param rawData The buffer to parse.
   * @param metadata The metadata of the document.
   * @returns Promise that resolves with an array of `Context` objects.
   */
  protected abstract parse(
    rawData: Buffer,
    metadata: Context['metadata']
  ): Promise<Context[]>;

  static async imports(): Promise<{ readFile: typeof readFile }> {
    try {
      const { readFile } = await import('node:fs/promises');

      return { readFile };
    } catch (e) {
      console.error(e);
      throw new Error('Failed to load fs/promises.');
    }
  }

  /**
   * Async method that reads the buffer contents and metadata, and then
   * calls the `parse()` method to parse the buffer and return the documents.
   * @returns Promise that resolves with an array of `Context` objects.
   */
  public async load(filePathOrBlob: CallInput): Promise<Context[]> {
    let buffer: Buffer;
    let metadata: Context['metadata'];

    if (typeof filePathOrBlob === 'string') {
      const { readFile } = await BufferLoader.imports();
      buffer = await readFile(filePathOrBlob);
      metadata = { source: filePathOrBlob };
    } else {
      buffer = await filePathOrBlob
        .arrayBuffer()
        .then((arrayBuffer) => Buffer.from(arrayBuffer));
      metadata = {
        source: 'blob',
        bolbType: filePathOrBlob.type,
        blobSize: filePathOrBlob.size,
      };
    }

    return this.parse(buffer, metadata);
  }
}
