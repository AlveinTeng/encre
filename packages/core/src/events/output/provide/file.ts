import fs from 'fs';
import { BaseSourceProvider } from './base.js';
import { GenerationChunk } from './generation.js';

export type GenerationFileChunkField = {
  /**
   * Generated file output.
   * @type {string} This output represents a file path.
   * @type {Blob} This output represetns a file in Blob.
   */
  output: string | Blob;

  info?: Record<string, unknown>;
};

export class GenerationFileChunk extends GenerationChunk {
  declare output: string | Blob;

  constructor(fields: GenerationFileChunkField) {
    super(fields);
  }

  protected _getOutputBlob(): Blob {
    if (typeof this.output === 'string') {
      return new Blob([fs.readFileSync(this.output)], {
        type: 'application/pdf',
      });
    }

    return this.output;
  }

  concat(chunk: GenerationFileChunk): GenerationFileChunk {
    const newBlob = new Blob([this._getOutputBlob(), chunk._getOutputBlob()], {
      type: 'application/pdf',
    });

    return new GenerationFileChunk({
      output: newBlob,
      info: {
        ...this.info,
        ...chunk.info,
        source: 'blob',
        type: newBlob.type,
        size: newBlob.size,
      },
    });
  }
}

/**
 * Class that provide file source as a file path or a file Blob.
 */
export class FileProvider extends BaseSourceProvider {
  declare source: string | Blob;

  declare generation: GenerationFileChunk;

  constructor(source: string | Blob) {
    super();
    this.source = source;
    this.generation = new GenerationFileChunk({
      output: source,
      info: {
        source: typeof source === 'string' ? source : 'blob',
        type: typeof source === 'string' ? undefined : source.type,
        size: typeof source === 'string' ? undefined : source.size,
      },
    });
  }

  replace(source: string | Blob) {
    this.generation = new GenerationFileChunk({
      output: source,
      info: {
        source: typeof source === 'string' ? source : 'blob',
        type: typeof source === 'string' ? undefined : source.type,
        size: typeof source === 'string' ? undefined : source.size,
      },
    });
  }

  add(source: string | Blob) {
    const newGeneration = new GenerationFileChunk({
      output: source,
      info: {
        source: typeof source === 'string' ? source : 'blob',
        type: typeof source === 'string' ? undefined : source.type,
        size: typeof source === 'string' ? undefined : source.size,
      },
    });

    this.generation.concat(newGeneration);
  }

  provide(): string | Blob {
    return this.generation.output;
  }
}
