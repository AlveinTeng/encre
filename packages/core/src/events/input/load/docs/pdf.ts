import { type Result, type Options, type Version } from 'pdf-parse';
import { Context } from '../context';
import { BufferLoader } from './buffer';

async function PDFLoaderImports() {
  try {
    const { default: PdfParse } = await import('pdf-parse');
    return { PdfParse };
  } catch (e) {
    console.error(e);
    throw new Error('Failed to load pdf-parse. Please install pdf-parse.');
  }
}

/**
 * Class that extends the `BufferLoader` class. It is a
 * document loader that loads documents from PDF files.
 */
export class PDFLoader<
  T extends string | Blob = string | Blob,
> extends BufferLoader<T> {
  private _splitPages: boolean;

  private _pdfjs: typeof PDFLoaderImports;

  constructor({ splitPages = true, pdfjs = PDFLoaderImports } = {}) {
    super();

    this._splitPages = splitPages;
    this._pdfjs = pdfjs;
  }

  /**
   * Method that taks a `rawData` buffer and `metadata` as parameters
   * and returns a promise that resolves to an array of `Context` instances.
   * @param rawData The buffer to be parsed.
   * @param metadata The metadata of the context.
   * @returns A promise that resolves to an array of `Context` instances.
   */
  public async parse(
    rawData: Buffer,
    metadata: Context['metadata']
  ): Promise<Context[]> {
    const { PdfParse } = await this._pdfjs();

    const pageContents: string[] = [];
    const renderPage = (pageData): string => {
      const renderOptions = {
        // replaces all occurrences of whitespace with standard spaces (0x20).
        // The default value is `false`.
        normalizeWhitespace: true,
        // do not attempt to combine same line TextItem's. The default value
        // is `false`.
        disableCombineTextItems: false,
      };

      return pageData
        .getTextContent(renderOptions)
        .then(function (textContent) {
          const text = textContent.items.map((item) => item.str).join('\n');
          pageContents.push(text);

          return text;
        });
    };

    const options: Options = {
      pagerender: renderPage,
      version: 'v1.10.100',
    };

    const pdf: Result = await PdfParse(rawData, options);
    const meta: any = pdf.metadata;
    const info: any = pdf.info;
    const version: Version = pdf.version;
    const numPages: number = pdf.numpages;

    const PdfDocuments: Context[] = [];
    for (let i = 1; i <= numPages; i++) {
      PdfDocuments.push(
        new Context({
          pageContent: pageContents[i - 1],
          metadata: {
            ...metadata,
            pdf: {
              version,
              info: info,
              metadata: meta,
              totalPages: numPages,
            },
            loc: {
              pageNumber: i,
            },
          },
        })
      );
    }

    if (this._splitPages) {
      return PdfDocuments;
    }

    if (PdfDocuments.length === 0) {
      return [];
    }

    return [
      new Context({
        pageContent: PdfDocuments.map((doc: Context) => doc.pageContent).join(
          '\n\n'
        ),
        metadata: {
          ...metadata,
          pdf: {
            version,
            info: info,
            metadata: meta,
            totalPages: numPages,
          },
        },
      }),
    ];
  }
}
