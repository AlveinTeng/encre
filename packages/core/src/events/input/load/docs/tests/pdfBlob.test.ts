import { Blob } from 'buffer';
import fs from 'fs';
import path from 'path';
import url from 'url';
import { expect, test } from '@jest/globals';
import { stringify } from 'yaml';
import { FileProvider } from '../../../../output/provide/file';
import { Context } from '../context';
import { PDFLoader } from '../pdf';

test('test PDF loader from blob', async () => {
  const filePath: string = path.resolve(
    path.dirname(url.fileURLToPath(import.meta.url)),
    './examples/NIPS-2014-generative-adversarial-nets-Paper.pdf'
  );

  const provider = new FileProvider(
    new Blob([fs.readFileSync(filePath)], { type: 'application/pdf' })
  );
  const loader = new PDFLoader();

  const serializedStr: string = JSON.stringify(loader, null, 2);
  expect(stringify(JSON.parse(serializedStr))).toMatchSnapshot();

  const docs: Context[] = await loader.load(provider.provide());

  expect(docs.length).toBe(9);
  expect(docs[0].pageContent).toContain('Generative Adversarial Nets');
  expect(stringify(docs[0].metadata)).toMatchSnapshot();

  expect(await loader.invoke(provider.provide())).toStrictEqual(docs);
});
