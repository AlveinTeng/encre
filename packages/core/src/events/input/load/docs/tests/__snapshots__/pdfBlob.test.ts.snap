// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test PDF loader from blob 1`] = `
"_grp: 1
_type: constructor
_id:
  - events
  - input
  - load
  - docs
  - pdf
  - PDFLoader
_kwargs: {}
"
`;

exports[`test PDF loader from blob 2`] = `
"source: blob
bolbType: application/pdf
blobSize: 539761
pdf:
  version: 1.10.100
  info:
    PDFFormatVersion: "1.3"
    IsAcroFormPresent: false
    IsXFAPresent: false
    Title: Generative Adversarial Nets
    Author: Ian Goodfellow, Jean Pouget-Abadie, Mehdi Mirza, Bing Xu, David
      Warde-Farley, Sherjil Ozair, Aaron Courville, Yoshua Bengio
    Subject: Neural Information Processing Systems http://nips.cc/
    Producer: PyPDF2
    ModDate: D:20141202174320-08'00'
  metadata: null
  totalPages: 9
loc:
  pageNumber: 1
"
`;
